### PR TITLE
Update some impl IR APIs to more modern alternatives

### PR DESCRIPTION
**Pull request by ZacSweers**
Just a little cleanup to use more stable IR factory functions rather than direct impl types

### CODE REVIEW

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt [hunk 1]
```diff
@@ -253,30 +259,20 @@ internal class ModuleMergerIr(
...
         putValueArgument(
           index = 0,
-          valueArgument = IrVarargImpl(
-            startOffset = UNDEFINED_OFFSET,
-            endOffset = UNDEFINED_OFFSET,
-            type = pluginContext.symbols.array.defaultType,
-            varargElementType = IrDynamicTypeImpl(null, emptyList(), INVARIANT),
-            elements = contributedModules
+          valueArgument = irVararg(
+            elementType = pluginContext.irBuiltIns.kClassClass.starProjectedType,
```
Comment (ZacSweers) [inline compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:272]: This is a change from before, but I think this is more correct (`Array<KClass<*>>`) as `dynamic` is only really intended for JS targets iirc

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt [hunk 2]
```diff
@@ -53,15 +53,21 @@ internal class ModuleMergerIr(
             .findAll(mergeComponentFqName, mergeSubcomponentFqName, mergeModulesFqName)
             .ifEmpty { return super.visitClass(declaration) }
 
-          generateDaggerAnnotation(annotations, moduleFragment, pluginContext, declarationReference)
+          pluginContext.irBuiltIns.createIrBuilder(declaration.symbol)
+            .generateDaggerAnnotation(
```
Comment (ZacSweers) [inline compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:57]: Creating an IR builder exposes more factory extension functions, including some of the ones used below

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt [hunk 3]
```diff
@@ -291,12 +287,11 @@ internal class ModuleMergerIr(
...
-              startOffset = 0,
-              endOffset = 0,
-              type = varargArguments[0].type,
-              varargElementType = varargArguments[0].varargElementType,
-              elements = varargArguments.flatMap { it.elements }
+            valueArgument = irVararg(
+              elementType = varargArguments[0].varargElementType,
+              // These are always IrExpression instances too
+              values = varargArguments.flatMap { it.elements }
+                .filterIsInstance<IrExpression>()
```
Comment (JoelWilcox) [inline compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:294]: Are we sure this is guaranteed? Wondering why the extension function enforces `IrExpressions` when the impl class doesn't

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt (hunk 1 above)
Comment (JoelWilcox) [reply compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:272]: What do you mean by this being a change from before?

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt (hunk 2 above)
Comment (JoelWilcox) [reply compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:57]: Do we need to add any followup `build()` equivalent for the IrBuilder? Seems like the answer is no but just wanted to verify

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt (hunk 3 above)
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:294]: Yeah if you look at the impl that's all it allows. My guess is it's a bit of a marker interface tomfoolery

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt (hunk 1 above)
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:272]: Before it was surprisingly the kotlin dynamic type, which I suspect was for lack of a better star alternative: https://kotlinlang.org/docs/dynamic-type.html

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt (hunk 2 above)
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:57]: Not as far as I know, it's more of a context thing than a true builder

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt (hunk 1 above)
Comment (JoelWilcox) [reply compiler/src/main/java/com/squareup/anvil/compiler/ModuleMergerIr.kt:272]: `The dynamic type is not supported in code targeting the JVM.` - mildly concerning :D, apparently it works though 😆. Good catch.

**Review by JoelWilcox — APPROVED** [review]

**Pull request merged**